💪V4

bluebill1049 opened the pull request:
## 🔨 These are the breaking changes

- [x] ✨improve module **exports**:

(we will export all types as well)

```
import { useForm } from 'react-hook-form';
``` 

🚨`import useForm from 'react-hook-form';`  will no longer work

---

- [x] ✨nested `errors` object and better typescript support

```
type form = {
  yourDetail: {
    firstName: string,
  }
}

errors?.yourDetail?.firstName
```

🚨no more flat errors :-) better TS support

---

- [x] ✨triggerValidation argument change from `Object`, `Object[]` to `String`, `String[]`

```
triggerValidation('firstName');
triggerValidation(['firstName', 'lastName']);
```
🚨`triggerValidation({ name: 'test', value: 'test'  });` no longer work

---

- [x] ✨watch support `{ nest: boolean }`

```
watch(); // { 'test.firstName': 'bill' }
watch({ nest: true }); // { test: { firstName: 'bill' } }
```

--- 

- [x] ✨improve custom `register`


```
register('test', { required: true });
```

🚨`register({ name: 'test'})` => `register('test')`

---
- [x] ✨`setError` support nested object

```
setError('yourDetail.firstName', 'test');
errors.yourDetails.firstName;
```

🚨`setError('yourDetail.firstName')` no longer output flat errors object 

---

## 🥂 New feature

- [x] `control`:  this will be the prop pass to RHFInput

we probably can skip this hack too: https://github.com/react-hook-form/react-hook-form/pull/635

```
<RHF control={control} name="test" />
```

https://github.com/react-hook-form/react-hook-form-input/pull/45

---

- [x] move `RHFInput` into the main repo and rename it to `Controller`

https://github.com/react-hook-form/react-hook-form-input

---

## ✂️ API to be removed list:

- [x] `validationSchemaOption`: hardly anyone want to use validation with abort early, or change the config.

---

- [x] native validation: hardly anyone used this feature. https://react-hook-form.com/api/#Browserbuiltinvalidation

FYI @JeromeDeLeon @kotarella1110 @adhrinae @krnlde 

Feedback welcome guys

bluebill1049 committed:
improve export modules

bluebill1049 committed:
fix unit test due to export

bluebill1049 committed:
convert touched from array to set

bluebill1049 committed:
remove extra functionality from triggerValidation

bluebill1049 committed:
fix unit test for triggerValidation

bluebill1049 committed:
working on nested errors object

JeromeDeLeon commented on src/useForm.ts:
```diff
@@ -635,6 +615,7 @@ export default function useForm<FormValues extends FieldValues = FieldValues>({
   }
 
   function watch(): FormValues;
+  function watch(option: { nest: boolean }): FormValues;
```
so if we specify `nest` to `true`, we aren't allowed to specify fields?

bluebill1049 replied:
not with `nest`, because you are getting an object and you can deconstruct it.

bluebill1049 replied:
or perhaps we can :)  but it will increase the complexity of the API. We can go there when there are people requesting it.

JeromeDeLeon replied:
I haven't use `watch` that much and therefore, I thought that the return would automatically  be constructed like `getValues({ nest: true })`

bluebill1049 replied:
yes you are right. it's basically `getValues`.

bluebill1049 committed:
convert more flat error into nest object

bluebill1049 committed:
working formstate automation fix due to touched

bluebill1049 commented:
<img width="506" alt="Screen Shot 2019-12-14 at 9 41 08 pm" src="https://user-images.githubusercontent.com/10513364/70847615-78f08380-1eba-11ea-9231-5a293f724fbd.png">

OMG! 666

![giphy](https://user-images.githubusercontent.com/10513364/70847631-99b8d900-1eba-11ea-815a-c02822afc26b.gif)

JeromeDeLeon commented:
> <img alt="Screen Shot 2019-12-14 at 9 41 08 pm" width="506" src="https://user-images.githubusercontent.com/10513364/70847615-78f08380-1eba-11ea-9231-5a293f724fbd.png">
> 
> OMG! 666
> 
> ![giphy](https://user-images.githubusercontent.com/10513364/70847631-99b8d900-1eba-11ea-815a-c02822afc26b.gif)

We're doing breaking changes, hence the number lol

bluebill1049 committed:
simplify register method and TS support

bluebill1049 committed:
simplify register method

bluebill1049 committed:
fix register with support name and multiple attribute

bluebill1049 committed:
only support string name for register

bluebill1049 committed:
fix ts for register method

bopfer commented:
I like all these changes and I think they are nice quality of life improvements.

One other potential change I would like to see, if possible, is related to Yup schema validation.  Currently, the validation returns the last error from the Yup validations. See this sandbox: https://codesandbox.io/s/react-hook-form-reset-form-2wg09

But, I think it would make more sense to return the first one, `required` in this case.  With the current `react-hook-form`, I need to invert all my Yup schemas to put the first things I want validated at the end.  However, I would prefer to have them at the start.  It's a minor issue and would be a breaking change for current Yup users, but it would be nice.

PS: I have tried using `validationSchemaOption: { abortEarly: true }`, but that doesn't seem to return any errors.  So, maybe there is just a bug there that needs to be fixed.  I am not really sure.  Either way, I think returning the first error would still be a better option.

bluebill1049 commented:
> But, I think it would make more sense to return the first one, `required` in this case. With the current `react-hook-form`, I need to invert all my Yup schemas to put the first things I want validated at the end. However, I would prefer to have them at the start. It's a minor issue and would be a breaking change for current Yup users, but it would be nice.

I think it totally make sense to return the first error it encounter, is this coming from Yup tho? is there any configure we can change to let Yup return the first encounter error.

bluebill1049 committed:
merge errors working with nested

JeromeDeLeon commented:
@bopfer Upon looking your concern, it actually return an error when using `validate` of `Yup`. The problem is when [parsingErrorSchema](https://github.com/react-hook-form/react-hook-form/blob/1bb6e543237c9b0f9a8e0e657bdd600912980061/src/logic/validateWithSchema.ts#L21). I'll try to look more into it.

bluebill1049 commented:
> @bopfer Upon looking your concern, it actually return an error when using `validate` of `Yup`. The problem is when [parsingErrorSchema](https://github.com/react-hook-form/react-hook-form/blob/1bb6e543237c9b0f9a8e0e657bdd600912980061/src/logic/validateWithSchema.ts#L21). I'll try to look more into it.

oh cool @JeromeDeLeon feel free to branch off from this PR to get this one right 👍

bluebill1049 committed:
remove validationSchemaOption

bluebill1049 committed:
fix unit test

bluebill1049 committed:
Merge branch 'master' into v4

JeromeDeLeon commented:
Oops. I re-added it again because of `abortEarly`

bluebill1049 commented:
> Oops. I re-added it again because of `abortEarly`

no worries, i think this validation schema option is pretty useless, hence removing it.

bluebill1049 commented:
just branch off from v4 branch if you want to change something @JeromeDeLeon

bluebill1049 committed:
reading errors using get method instead object access

JeromeDeLeon commented:
@bopfer , Seems like we're removing the `abortEarly` for the next version. I tried searching for use case of `abortEarly` and based on this [issue](https://github.com/jquense/yup/issues/718), the default value of `abortEaly` is `true` according to [Yup](https://github.com/jquense/yup#mixedvalidatevalue-any-options-object-promiseany-validationerror), they run in parallel and the first error the emit will be returned to you. It's not the first validation you set per field but the first error the comes out when you validate.

bluebill1049 commented:
> @bopfer , Seems like we're removing the `abortEarly` for the next version. I tried searching for use case of `abortEarly` and based on this [issue](https://github.com/jquense/yup/issues/718), the default value of `abortEaly` is `true` according to [Yup](https://github.com/jquense/yup#mixedvalidatevalue-any-options-object-promiseany-validationerror), they run in parallel and the first error the emit will be returned to you. It's not the first validation you set per field but the first error the comes out when you validate.

my understanding `abortEarly: true` is whenever the first error encounter, and stop rest of the validation.

what @bopfer is after is order of the validation result.

```
const validationSchema = object().shape({
  firstName: string()
    .required()
    .min(3)
});
```

this will result `firstName must be at least 3 characters` first appear than required, i think it's our error parsing or the order of yup validation. (mostly due to us, i haven't look close enough).

bopfer commented:
> @bopfer , Seems like we're removing the `abortEarly` for the next version. I tried searching for use case of `abortEarly` and based on this [issue](https://github.com/jquense/yup/issues/718), the default value of `abortEaly` is `true` according to [Yup](https://github.com/jquense/yup#mixedvalidatevalue-any-options-object-promiseany-validationerror), they run in parallel and the first error that emitted will be returned to you. It's not the first validation you set per field like `.required().min(3)` but the first error from the form the comes out when you validate.

I think removing `abortEarly` is fine.  It doesn't appear to work right now anyway.  `false` appears to be the default and `true` doesn't return any errors.

> I think it's our error parsing or the order of yup validation. (mostly due to us, i haven't look close enough).

It would be awesome if something can just be flipped in the parsing to return the first error.  Assuming Yup gives the errors an in order that makes sense.

JeromeDeLeon commented:
@bopfer let me know your thoughts about this #668

bluebill1049 committed:
working version with build-in submit nested errors object

JeromeDeLeon committed:
yup validation error appear order

* Support abort errors early

* Fix linter

bluebill1049 committed:
unset method to remove deep error object

bluebill1049 committed:
support multiple reset

bluebill1049 committed:
unset error message instead of omit

bluebill1049 committed:
fix validation schema snapshot

JeromeDeLeon commented on cypress/integration/formState.ts:
```diff
@@ -3,38 +3,38 @@ context('form state', () => {
 
     cy.get('#state').contains(
-      '{"dirty":false,"isSubmitted":false,"submitCount":0,"touched":[],"isSubmitting":false,"isValid":false}',
+      '{"dirty":false,"isSubmitted":false,"submitCount":0,"touched":{},"isSubmitting":false,"isValid":false}',
```
This is really clever! So this is how you get a hold of component state by printing them on the dom and track them. Nice!

JeromeDeLeon replied:
This is what I wanted. I mean, I really don't know I to monitor my state in testing. So this is one way of doing it. :)

bluebill1049 committed:
attach array example

bopfer commented:
> @bopfer let me know your thoughts about this #668

That looks like it will take care of what I wanted.  Thanks!

bluebill1049 commented:
> > @bopfer let me know your thoughts about this #668
> 
> That looks like it will take care of what I wanted. Thanks!

by the way, we are not fixing it for @bopfer. We think that's the right and consistent way for the validation report. It's more of a correction.

bluebill1049 committed:
introduce control method for controlled component

bluebill1049 committed:
working on new control API

krnlde commented:
Oh wow, I just come out of the weekend and have so much to read :D Sorry guys I won't manage to pick up everything.

bluebill1049 commented:
> Oh wow, I just come out of the weekend and have so much to read :D Sorry guys I won't manage to pick up everything.

thanks! that's all good, just take a look at the summary on the top. let us know your feedback. much appreciated.

kotarella1110 commented on src/useForm.ts:
```diff
@@ -323,9 +304,7 @@ export function useForm<FormValues extends FieldValues = FieldValues>({
-      const names = isArray(payload)
-        ? payload.map(({ name }) => name)
-        : [payload.name];
+      const names = isArray(payload) ? payload.map(name => name) : [name];
```
`map` is unnecessary.
Isn't it a payload `instead` of a `name`?

the following is correct:

```ts
const names = isArray(payload) ? payload : [payload];
```

bluebill1049 replied:
Awesome pick up. Send PR to this branch if u want ❤️

kotarella1110 replied:
Thanks. I will!

kotarella1110 replied:
Sorry, this is already fixed :scream:

kotarella1110 commented on src/useForm.ts:
```diff
@@ -363,13 +342,11 @@ export function useForm<FormValues extends FieldValues = FieldValues>({
     ): Promise<boolean> => {
       const fields =
-        payload || Object.keys(fieldsRef.current).map(name => ({ name }));
+        payload || Object.keys(fieldsRef.current).map(name => name);
```
`map` is unnecessary.

the following is correct:
```ts
const fields = payload || Object.keys(fieldsRef.current);
```

bluebill1049 replied:
Correct!

bluebill1049 committed:
remove mergeErrors and remove debugger in code

bluebill1049 committed:
useCallback on control method and unset error instead of omit

kotarella1110 committed:
remove unnecessary map function within triggerValidation (#679)

bluebill1049 committed:
remove native validation

bluebill1049 committed:
TS support for nested error object

bluebill1049 commented on src/types.ts:
```diff
@@ -103,9 +103,15 @@ export type FieldRefs<FormValues extends FieldValues> = Partial<
-  Record<FieldName<FormValues>, FieldError>
->;
+export type FieldErrors<FormValues> = {
+  [Key in keyof FormValues]?: FormValues[Key] extends any[]
```
@stramel review here plz

stramel replied:
This looks fine to me. Looks pretty similar to the other lib's typing that we were discussing.

stramel replied:
**NOTE:** I didn't test this

bluebill1049 replied:
awesome thanks @stramel i did test it.

krnlde commented:
I read the summary and I really like the new stuff. My only bit of concern would be the Set instead of Array regarding browser compatibility. But if the pipeline already transpiles/polyfills that everything's cool :) This'll be a major version bump, right?

bluebill1049 commented:
> I read the summary and I really like the new stuff. My only bit of concern would be the Set instead of Array regarding browser compatibility. But if the pipeline already transpiles/polyfills that everything's cool :)

good point. alternative we can convert into `Object`? 

1. `touched.test`

2. `touched.includes('test')`

3. `touched.has('test')`

@kotarella1110 @JeromeDeLeon @bopfer @krnlde which one is you guys reckon is easier for developers.

bluebill1049 committed:
change control to props and pass automation

bluebill1049 committed:
Merge branch 'master' into v4

bluebill1049 commented:
Early beta version: `4.0.0-beta.1`

bluebill1049 committed:
expose more methods and information about RHF with RHFInput

Pratikkakkad commented:
when can we expect this to release? I am just really excited 😄

bluebill1049 commented:
> when can we expect this to release? I am just really excited 😄

Thanks @Pratikkakkad I am still working on it. I will ping you once we are ready or have release candidate version.

bluebill1049 committed:
bring RHFInput into the main repo

bluebill1049 committed:
clean up RHFInput

bluebill1049 committed:
clean up RHFInput logic

bluebill1049 committed:
- Improve RHFInput
- Remove utils index.ts
- Improve context types
- Update readme
- Include new function to detect validation

bluebill1049 committed:
revert back on how context type was defined :(

bluebill1049 committed:
include automation test for RHFInput and add test for skipValidation

bluebill1049 committed:
fix a bug discovered by automation

bluebill1049 committed:
rename RHFInput to Controller

bluebill1049 committed:
update automation with error messages

JeromeDeLeon commented on app/src/basic.tsx:
```diff
@@ -20,15 +20,35 @@ const Basic: React.FC = (props: any) => {
 
   renderCounter++;
 
+  console.log(errors);
```
Are you doing your testing here? @bluebill1049

bluebill1049 replied:
oops i will remove that!

bluebill1049 committed:
start working on unit test for Controller and automation

bluebill1049 commented:
`react-hook-form@4.0.0-beta.2`

bluebill1049 committed:
adding unit tests for controller and upgrade packages

bluebill1049 committed:
add unit test for getInputValue

bluebill1049 committed:
upgrade automation app packages

bluebill1049 committed:
revert package uninstall

bluebill1049 committed:
clean up type and remove NameProp

bluebill1049 committed:
add another test suit for controller default value

bluebill1049 committed:
4.0.0-rc.0

bluebill1049 committed:
4.0.0-rc.1

bluebill1049 committed:
Merge branch 'master' into v4

bluebill1049 commented:
Hey guys, if you have free sometimes. Please helping on testing the new RC: `react-hook-form@4.0.0-rc.1`

bluebill1049 committed:
consistent with useCallback

bluebill1049 committed:
Merge remote-tracking branch 'origin/v4' into v4

bluebill1049 committed:
fix an issue with Controller Component reset

bluebill1049 committed:
fix issue with setValue

bluebill1049 committed:
fix unit test for controller

bluebill1049 committed:
fix lint issue

bluebill1049 committed:
4.0.0-rc.2

bluebill1049 commented:
`react-hook-form@4.0.0-rc.2`

bluebill1049 committed:
update example.gif

bluebill1049 committed:
update example.gif

bluebill1049 committed:
update all the readme code example

bluebill1049 committed:
update all example import for useForm

bluebill1049 committed:
close #691 filter undefined/null in array data

bluebill1049 committed:
4.0.0-rc.3

bluebill1049 committed:
only filter out undefined and null

bluebill1049 commented:
`react-hook-form@4.0.0-rc.3`

bluebill1049 commented:
Looks pretty good at my end:

https://codesandbox.io/s/react-hook-form-v4-quick-test-9e18f

bluebill1049 committed:
clean up code

bluebill1049 committed:
convert example single quote to double

bluebill1049 committed:
prettier all the examples

bluebill1049 committed:
remove all import css

bluebill1049 committed:
update example and simplify parseErrorSchema logic

bluebill1049 committed:
remove isObjectString method

avocadowastaken committed:
feat: Expose `contextTypes`. (#695)

avocadowastaken committed:
feat: Add `control` to `FormContextValues`, return `FormContextValues` in `useForm`. (#694)

bluebill1049 commented:
Thanks again for your help @umidbekkarimov 🙏

bluebill1049 commented:
Thanks 💪@JeromeDeLeon 💪@kotarella1110 and 💪@umidbekkarimov 

merged ❤️

AhmedAymanM commented on src/logic/skipValidation.test.ts:
```diff
@@ -0,0 +1,131 @@
+        isOnSubmit: false,
+        isReValidateOnSubmit: true,
+        isOnBlur: false,
+        isReValidateOnBlur: false,
```
what I get from the test message is that `isReValidateOnSubmit` should be `false` and `isReValidateOnBlur` should be `true`

bluebill1049 replied:
you are right! feel free to send a PR 🙏

AhmedAymanM replied:
opened a PR :) 
https://github.com/react-hook-form/react-hook-form/pull/702
best of luck 🙏

bluebill1049 replied:
merged thank you :)

AhmedAymanM replied:
awesome! there is no need! :)

babramczyk commented:
@bluebill1049 Awesome work! Thanks for the nested `errors` object, it's a huge help!

bluebill1049 commented:
> @bluebill1049 Awesome work! Thanks for the nested `errors` object, it's a huge help!

nice! i am glad you find it useful 👍 now that optional chaining is adopted by most of the transpilers.
